File: board/web/test/register-taxonomy.test.js
// register-taxonomy.test.js - pins the register taxonomy's unpinned edges (#37).
//
// EDGE 1 (JS): compose.js REGISTER_ORDER vs schema/yard-snapshot.schema.json
// $defs.register.enum.
//
// This test reads the real schema file from disk (same shape as the
// sse-event-name.test.js exemplar: board/web/test -> board/web -> board ->
// repo root, then schema/). It pins SET EQUALITY - every value in REGISTER_ORDER
// must appear in the schema enum and vice versa.
//
// ORDERING IS COMPOSE.JS-OWNED: the schema enum cannot express severity order
// (JSON Schema enums are unordered); mostSevereRegister (compose.js:23-25)
// depends on REGISTER_ORDER's array index position. The schema enum is the
// CONTENT authority; compose.js is the ORDER authority. This test enforces
// content agreement and explicitly does NOT order-check the schema. #37.
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { REGISTER_ORDER } from '../js/compose.js';

const thisDir = dirname(fileURLToPath(import.meta.url));
const repoRoot = dirname(dirname(dirname(thisDir))); // board/web/test -> board/web -> board -> repo root
const schema = JSON.parse(readFileSync(join(repoRoot, 'schema', 'yard-snapshot.schema.json'), 'utf8'));

test('REGISTER_ORDER and schema $defs.register.enum are SET-EQUAL (content authority: schema; order authority: compose.js) #37', () => {
  const schemaEnum = schema.$defs.register.enum;
  assert.ok(
    Array.isArray(schemaEnum),
    'schema.$defs.register.enum must be an array (fixture sanity)'
  );

  // Non-vacuity: both sides must be non-empty.
  assert.ok(REGISTER_ORDER.length > 0, 'REGISTER_ORDER must be non-empty');
  assert.ok(schemaEnum.length > 0, 'schema.$defs.register.enum must be non-empty');

  // Every REGISTER_ORDER value must be in the schema enum.
  for (const r of REGISTER_ORDER) {
    assert.ok(
      schemaEnum.includes(r),
      `REGISTER_ORDER value '${r}' is not in schema.$defs.register.enum - schema and compose.js have drifted (#37)`
    );
  }

  // Every schema enum value must be in REGISTER_ORDER.
  for (const r of schemaEnum) {
    assert.ok(
      REGISTER_ORDER.includes(r),
      `schema.$defs.register.enum value '${r}' is not in REGISTER_ORDER - schema and compose.js have drifted (#37)`
    );
  }

  // Set sizes must match (catches duplicates on either side).
  assert.equal(
    REGISTER_ORDER.length,
    schemaEnum.length,
    `REGISTER_ORDER has ${REGISTER_ORDER.length} values but schema enum has ${schemaEnum.length} - sets differ (#37)`
  );
});

File: board/assemble/register_taxonomy_test.go
package assemble

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// TestClosedRegistersMatchSchemaEnum pins EDGE 2 of the register taxonomy (#37):
// boarddefs.go's closedRegisters map vs schema/yard-snapshot.schema.json
// $defs.register.enum. Reads the real schema from disk using the repoRoot helper
// (testroot_test.go in this package), the same robustness argument as every
// other schema-reading test in this repo.
//
// ORDERING IS COMPOSE.JS-OWNED (schema/JS boundary): the schema enum cannot
// express severity order (JSON Schema enums are unordered); mostSevereRegister
// in compose.js depends on REGISTER_ORDER's array-index position. This test
// enforces SET EQUALITY only; order is not checked here. #37.
//
// Fault-injection evidence (recorded in commit message): temporarily adding
// "fault-injection-probe" to closedRegisters caused this test to FAIL with
// "closedRegisters key \"fault-injection-probe\" not in schema.$defs.register.enum".
// Reverted byte-identical before commit.
func TestClosedRegistersMatchSchemaEnum(t *testing.T) {
	schemaPath := filepath.Join(repoRoot(t), "schema", "yard-snapshot.schema.json")
	data, err := os.ReadFile(schemaPath)
	if err != nil {
		t.Fatalf("could not read %s: %v", schemaPath, err)
	}

	// Parse only the $defs.register.enum portion of the schema.
	// The `json:"$defs"` tag is valid Go (json tags are arbitrary strings).
	var sc struct {
		Defs struct {
			Register struct {
				Enum []string `json:"enum"`
			} `json:"register"`
		} `json:"$defs"`
	}
	if err := json.Unmarshal(data, &sc); err != nil {
		t.Fatalf("could not parse schema: %v", err)
	}

	schemaEnum := sc.Defs.Register.Enum

	// Non-vacuity guards: both sides must be non-empty.
	if len(schemaEnum) == 0 {
		t.Fatal("schema.$defs.register.enum is empty - fixture sanity check failed")
	}
	if len(closedRegisters) == 0 {
		t.Fatal("closedRegisters is empty - fixture sanity check failed")
	}

	// Every closedRegisters key must appear in the schema enum.
	for r := range closedRegisters {
		found := false
		for _, s := range schemaEnum {
			if s == r {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("closedRegisters key %q is not in schema.$defs.register.enum - boarddefs.go and schema have drifted (#37)", r)
		}
	}

	// Every schema enum value must appear in closedRegisters.
	for _, s := range schemaEnum {
		if !closedRegisters[s] {
			t.Errorf("schema.$defs.register.enum value %q is not in closedRegisters - boarddefs.go and schema have drifted (#37)", s)
		}
	}

	// Size equality catches duplicates on either side.
	if len(closedRegisters) != len(schemaEnum) {
		t.Errorf("closedRegisters has %d entries but schema enum has %d - sets differ (#37)", len(closedRegisters), len(schemaEnum))
	}
}
